faculty/clients/notification.py

```diff
@@ -0,0 +1,102 @@
+# Copyright 2020 Faculty Science Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Consume notifications from the Faculty frontend.
+"""
+
+import json
+
+import sseclient
+
+from faculty.clients.base import BaseClient
+
+
+class NotificationClient(BaseClient):
+    """Client to listen on notifications from the Faculty frontend.
+
+    Either build this client with a session directly, or use the
+    :func:`faculty.client` helper function:
+
+    >>> client = faculty.client("notification")
+
+    Parameters
+    ----------
+    session : faculty.session.Session
+        The session to use to make requests
+    """
+
+    _SERVICE_NAME = "frontend"
+
+    def user_updates(self, user_id):
+        """Get notification events for the given user.
+
+        Parameters
+        ----------
+        user_id : uuid.UUID
+            ID of the user to get updates for.
+
+        Returns
+        -------
+        generator
+            Server-sent events
+        """
+        endpoint = "api/updates/user/{}".format(user_id)
+        response = self._get_raw(endpoint, stream=True)
+        client = sseclient.SSEClient(response)
+        return client.events()
+
+    def check_publish_template_result(self, events, project_id):
```

Comment:
Are we assuming that only one template is published at a time on a project? What if two users publish templates from the same project at similar times?

Reply:
We assume 1 template is published at the same time for 1 (project, user).

The `event` passed to `check_publish_template_result` are meant to be those returned by `user_updates` which are for 1 user.

Yes, this could break if 1 user published 2 templates at once. I think frontend would break in the same way, it also just listens to the (project+user) updates. I considered handling this by generating a unique request ID for later checks, but then decided it wasn't worth the hassle given how likely I think this is going to be used and how likely it is for 1 user to run publishing more than once at a time.

Reply:
Hmm, ok.